doc/lenses.rst

```diff
@@ -3,16 +3,494 @@
 Lenses
 ======
 
-.. admonition:: Work in progress
-   :class: danger
+A tool borrowed from functional programming. While lenses are used
+extensively in Lager for zooming on cursors, they are also useful on
+their own for manipulating immutable data.
 
-   This section of the documentation has not been written yet.
+.. _making-a-lens:
 
-   Meanwhile, you can watch the **C++ Russia 2019 Talk**: `Squaring
-   the circle: value oriented design in an object oriented system
-   <https://www.youtube.com/watch?v=e2-FRFEx8CA>`_ (`slides`_).
+Making a Lens
+-------------
 
-.. _slides: https://sinusoid.es/talks/cpprussia19-piter
+Lenses are, conceptually, a pair of functions for focusing on a part
+of a whole. You use a lens with the following interface:
 
-.. image:: http://orvex.org.p11.hostingprod.com/images/under_construction.gif
+.. code-block:: c++
+
+   // pseudocode for the lens interface:
+   Part view(lens<Whole, Part>, Whole); // get part of a whole
+   Whole set(lens<Whole, Part>, Whole, Part); // set the part of a whole
+   Whole over(lens<Whole, Part>, Whole, std::function<Part(Part)>); // update the part of a whole
+   lens<A, C> compose(lens<A, B>, lens<B, C>); // compose two lenses into another lens
+
+Let's use this mouse as our model:
+
+.. image:: _static/mouse.svg
    :align: center
+
+.. code-block:: c++
+
+   struct Mouse {
```

Comment:
I've notived you have used `CamelCase` and `camelCase` throughout the file. I'd suggest using `snake_case` for consistency with the rest of the documentation.

Reply:
I changed everything to use snake case, but I'm not sure if naming struct members with the same name as types is acceptable or not.

Reply:
Clang normally accepts it but GCC does not (I think the standard does not allow it actually.)

Reply:
how should we fix this then?

Reply:
Hmmm. You can use `_t` for types. But I noticed there are many cases here, maybe I am ok to accept `CamelCase` for types in this page of the documentation if your prefer that style (it can be your personal signature ;) ). But in any case keep `snake_case` for members, functions and and variables.

Reply:
I think I'll go with PascalCase for classes and snake_case for the rest, I feel like it's better for readability.

Reply:
I decided on reverting to PascalCase for some class names.